File: frontend/apps/ios/Pantopus/Features/Scheduling/Setup/Wizard/FirstRunWizardScreen.swift
//
//  FirstRunWizardScreen.swift
//  Pantopus
//
//  A2 "Set up booking" first-run wizard screen. Wraps `WizardShell` with the
//  owner's identity and switches on `model.step` to render the StepRail + step
//  body (claim handle · pick type · weekly hours · success hero). Success share
//  presents a `UIActivityViewController` via `.sheet(item:)`; dismissing it
//  finishes the wizard. Matches `scheduling-setup-frames.jsx`.
//

import SwiftUI
import UIKit

struct FirstRunWizardScreen: View {
    @State private var model: FirstRunWizardModel
    @Environment(\.dismiss) private var dismiss

    init(owner: SchedulingOwner) {
        _model = State(wrappedValue: FirstRunWizardModel(owner: owner))
    }

    var body: some View {
        WizardShell(model: model, identity: model.theme.identity) {
            content
        }
        .navigationBarBackButtonHidden(true)
        .onChange(of: model.isFinished) { _, finished in
            if finished { dismiss() }
        }
        .sheet(item: shareBinding) { item in
            WizardShareSheet(items: [item.url])
        }
    }

    private var shareBinding: Binding<ShareItem?> {
        Binding(
            get: { model.pendingShareURL.map(ShareItem.init) },
            set: { newValue in
                if newValue == nil {
                    model.pendingShareURL = nil
                    model.finishAfterShare()
                }
            }
        )
    }

    private var steps: [(Int, String)] {
        [(1, "Link"), (2, "Type"), (3, "Hours"), (4, "Share")]
    }

    @ViewBuilder
    private var content: some View {
        WizardStepRail(steps: steps, current: model.step.rawValue, accent: model.theme.accent, accentBg: model.theme.accentBg)
        switch model.step {
        case .link:
            WizardHeadline(
                title: "Claim your booking link",
                sub: "This is the link you'll share. People book you at it — pick something short and memorable."
            )
            WizardHandleField(
                slug: $model.slug,
                state: model.slugState,
                accent: model.theme.accent,
                accentBg: model.theme.accentBg,
                onPick: { model.pickSuggestion($0) }
            )
        case .type:
            WizardHeadline(
                title: "Pick a meeting type",
                sub: "Start with one — how you meet and how long it runs. You can add more from settings."
            )
            WizardTypePicker(
                locationMode: $model.locationMode,
                duration: $model.duration,
                accent: model.theme.accent,
                accentBg: model.theme.accentBg
            )
        case .hours:
            WizardHeadline(
                title: "Set your weekly hours",
                sub: "People can only book inside these windows. You can fine-tune any day, or just use the defaults."
            )
            WizardTimezoneChip(identifier: model.timezoneIdentifier, accent: model.theme.accent, accentBg: model.theme.accentBg)
            WizardHoursGrid(enabled: $model.hoursEnabled, accent: model.theme.accent, rangeLabel: model.rangeLabel)
        case .success:
            WizardSuccessHero(
                accent: model.theme.accent,
                accentBg: model.theme.accentBg,
                shadow: model.theme.ctaShadow,
                title: "You're all set",
                sub: "Your booking link is live. Share it and people can book a \(model.duration)-minute meeting during your weekly hours.",
                link: model.shareLink,
                onCopy: copyLink
            )
        }
    }

    private func copyLink() {
        UIPasteboard.general.string = "https://\(model.shareLink)"
    }
}

// MARK: - Share sheet plumbing

private struct ShareItem: Identifiable {
    let url: URL
    var id: String { url.absoluteString }
}

private struct WizardShareSheet: UIViewControllerRepresentable {
    let items: [Any]

    func makeUIViewController(context _: Context) -> UIActivityViewController {
        UIActivityViewController(activityItems: items, applicationActivities: nil)
    }

    func updateUIViewController(_: UIActivityViewController, context _: Context) {}
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/scheduling/setup/FirstRunWizardScreen.kt
@file:Suppress("PackageNaming", "MagicNumber", "LongMethod", "UnusedParameter")

package app.pantopus.android.ui.screens.scheduling.setup

import android.content.Intent
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.platform.LocalContext
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import app.pantopus.android.ui.screens.scheduling._shared.SchedulingPillar
import app.pantopus.android.ui.screens.shared.wizard.WizardShell
import app.pantopus.android.ui.screens.shared.wizard.blocks.HeadlineBlock
import app.pantopus.android.ui.screens.shared.wizard.blocks.SubcopyBlock
import app.pantopus.android.ui.screens.shared.wizard.blocks.SuccessHeroBlock

@Composable
fun FirstRunWizardScreen(
    onBack: () -> Unit = {},
    onNavigate: (String) -> Unit = {},
    viewModel: FirstRunWizardViewModel = hiltViewModel(),
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val pendingShareUrl by viewModel.pendingShareUrl.collectAsStateWithLifecycle()
    val finished by viewModel.finished.collectAsStateWithLifecycle()
    val context = LocalContext.current
    val pillar = SchedulingPillar.Personal

    LaunchedEffect(finished) { if (finished) onBack() }

    LaunchedEffect(pendingShareUrl) {
        pendingShareUrl?.let { url ->
            val send =
                Intent(Intent.ACTION_SEND).apply {
                    type = "text/plain"
                    putExtra(Intent.EXTRA_TEXT, url)
                }
            runCatching { context.startActivity(Intent.createChooser(send, null)) }
            viewModel.shareConsumed()
        }
    }

    WizardShell(model = viewModel, identity = pillar.wizardIdentity) {
        WizardStepRail(steps = listOf("Link", "Type", "Hours", "Share"), current = state.step, pillar = pillar)
        when (state.step) {
            1 -> {
                HeadlineBlock("Claim your booking link")
                SubcopyBlock("This is the link you'll share. People book you at it — pick something short and memorable.")
                SlugClaimField(
                    overline = "Your link",
                    slug = state.slug,
                    state = state.slugState,
                    availableHint = "People will book you at this link.",
                    onSlugChange = viewModel::onSlugChange,
                    onPickSuggestion = viewModel::onPickSuggestion,
                    pillar = pillar,
                )
            }
            2 -> {
                HeadlineBlock("Pick a meeting type")
                SubcopyBlock("Start with one — how you meet and how long it runs. You can add more from settings.")
                WizardTypePicker(
                    selectedMode = state.locationMode,
                    duration = state.duration,
                    pillar = pillar,
                    onSelectMode = viewModel::onSelectLocation,
                    onSelectDuration = viewModel::onSelectDuration,
                )
            }
            3 -> {
                HeadlineBlock("Set your weekly hours")
                SubcopyBlock("People can only book inside these windows. You can fine-tune any day, or just use the defaults.")
                WizardTimezoneChip(timezoneId = state.timezoneId, pillar = pillar)
                WizardHoursGrid(hours = state.hours, pillar = pillar, onToggleDay = viewModel::onToggleDay)
            }
            else -> {
                SuccessHeroBlock(
                    headline = "You're all set",
                    subcopy =
                        "Your booking link is live. Share it and people can book a " +
                            "${state.duration}-minute meeting during your weekly hours.",
                )
                WizardSuccessLinkCard(
                    link = state.shareLink,
                    pillar = pillar,
                    onCopy = viewModel::onPrimary,
                )
            }
        }
    }
}
